### ISSUE TITLE
Random Server Freezes after updating to 1.14

### ISSUE THREAD

**Issue by dogboy21** [PaperMC/Paper#2450]
Since updating our server to 1.14 we suffer from random server freezing. The server doesn't even recover from it after a few minutes and needs to be killed.
Sometimes it even happens straight after starting.

### Link/paste of stack trace
I've collected a few thread dumps before killing the server:

<details>
    <summary>Thread Dump 1</summary>

    Server thread #25 prio=5 os_prio=0 tid=0x00007faaad444000 nid=0x6791 runnable [0x00007fa1e0fa3000]
        java.lang.Thread.State: TIMED_WAITING (parking)
            at sun.misc.Unsafe.park(Native Method)
            - parking to wait for  <0x00000002f186c8b0> (a java.lang.String)
            at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
            at net.minecraft.server.v1_14_R1.MinecraftServer.executeModerately(MinecraftServer.java:986)
            at net.minecraft.server.v1_14_R1.MinecraftServer.loadSpawn(MinecraftServer.java:608)
            at org.bukkit.craftbukkit.v1_14_R1.CraftServer.createWorld(CraftServer.java:1021)
            at org.bukkit.Bukkit.createWorld(Bukkit.java:492)
            at de.myftb.worldmanagement.WorldManagement.loadWorld(WorldManagement.java:352)
            at de.myftb.worldmanagement.WorldManagement.loadWorld(WorldManagement.java:290)
            at de.myftb.worldmanagement.listener.Join$3.run(Join.java:205)
            at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftTask.run(CraftTask.java:84)
            at org.bukkit.craftbukkit.v1_14_R1.scheduler.CraftScheduler.mainThreadHeartbeat(CraftScheduler.java:452)
            at net.minecraft.server.v1_14_R1.MinecraftServer.b(MinecraftServer.java:1150)
            at net.minecraft.server.v1_14_R1.DedicatedServer.b(DedicatedServer.java:417)
            at net.minecraft.server.v1_14_R1.MinecraftServer.a(MinecraftServer.java:1077)
            at net.minecraft.server.v1_14_R1.MinecraftServer.run(MinecraftServer.java:921)
            at java.lang.Thread.run(Thread.java:748) 
</details>

<details>
    <summary>Thread Dump 2</summary>

    Server thread #25 prio=5 os_prio=0 tid=0x00007f236943c000 nid=0x6241 waiting on condition [0x00007f1a9c807000]
        java.lang.Thread.State: TIMED_WAITING (parking)
            at sun.misc.Unsafe.park(Native Method)
            - parking to wait for  <0x00000002da6447e8> (a java.lang.String)
            at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
            at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.bi(IAsyncTaskHandler.java:131)
            at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:120)
            at net.minecraft.server.v1_14_R1.ChunkProviderServer.getChunkAt(ChunkProviderServer.java:154)
            at net.minecraft.server.v1_14_R1.World.getChunkAt(World.java:376)
            at net.minecraft.server.v1_14_R1.IWorldReader.getChunkAt(IWorldReader.java:80)
            at net.minecraft.server.v1_14_R1.World.getChunkAt(World.java:337)
            at net.minecraft.server.v1_14_R1.World.getType(World.java:729)
            at org.bukkit.craftbukkit.v1_14_R1.block.CraftBlock.getType(CraftBlock.java:211)
            at com.sk89q.worldguard.bukkit.listener.RegionFlagsListener.onPlaceBlock(RegionFlagsListener.java:68)
            at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor181.execute(Unknown Source)
            at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69)
            at org.bukkit.plugin.EventExecutor$$Lambda$2608/1120013847.execute(Unknown Source)
            at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
            at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
            at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:545)
            at com.sk89q.worldguard.bukkit.util.Events.fireToCancel(Events.java:74)
            at com.sk89q.worldguard.bukkit.listener.EventAbstractionListener.onEntityChangeBlock(EventAbstractionListener.java:287)
            at com.destroystokyo.paper.event.executor.asm.generated.GeneratedEventExecutor206.execute(Unknown Source)
            at org.bukkit.plugin.EventExecutor.lambda$create$1(EventExecutor.java:69)
            at org.bukkit.plugin.EventExecutor$$Lambda$2608/1120013847.execute(Unknown Source)
            at co.aikar.timings.TimedEventExecutor.execute(TimedEventExecutor.java:80)
            at org.bukkit.plugin.RegisteredListener.callEvent(RegisteredListener.java:70)
            at org.bukkit.plugin.SimplePluginManager.callEvent(SimplePluginManager.java:545)
            at org.bukkit.craftbukkit.v1_14_R1.event.CraftEventFactory.callEntityChangeBlockEvent(CraftEventFactory.java:1118)
            at org.bukkit.craftbukkit.v1_14_R1.event.CraftEventFactory.callEntityChangeBlockEvent(CraftEventFactory.java:1110)
            at net.minecraft.server.v1_14_R1.EntityFallingBlock.tick(EntityFallingBlock.java:140)
            at net.minecraft.server.v1_14_R1.WorldServer.entityJoinedWorld(WorldServer.java:584)
            at net.minecraft.server.v1_14_R1.WorldServer$$Lambda$3626/1245828625.accept(Unknown Source)
            at net.minecraft.server.v1_14_R1.World.a(World.java:942)
            at net.minecraft.server.v1_14_R1.WorldServer.doTick(WorldServer.java:356)
            at net.minecraft.server.v1_14_R1.MinecraftServer.b(MinecraftServer.java:1210)
            at net.minecraft.server.v1_14_R1.DedicatedServer.b(DedicatedServer.java:417)
            at net.minecraft.server.v1_14_R1.MinecraftServer.a(MinecraftServer.java:1077)
            at net.minecraft.server.v1_14_R1.MinecraftServer.run(MinecraftServer.java:921)
            at java.lang.Thread.run(Thread.java:748) 
</details>

<details>
    <summary>Thread Dump 3</summary>

    "Server thread" #27 prio=5 os_prio=0 tid=0x00007fcd7d075800 nid=0x7f2f runnable [0x00007fcd05e7e000]
        java.lang.Thread.State: TIMED_WAITING (parking)
            at sun.misc.Unsafe.park(Native Method)
            - parking to wait for  <0x00000002d2e4bb40> (a java.lang.String)
            at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:215)
            at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.bi(IAsyncTaskHandler.java:131)
            at net.minecraft.server.v1_14_R1.IAsyncTaskHandler.awaitTasks(IAsyncTaskHandler.java:120)
            at net.minecraft.server.v1_14_R1.ChunkProviderServer.getChunkAt(ChunkProviderServer.java:284)
            at net.minecraft.server.v1_14_R1.IChunkProvider.getChunkAt(SourceFile:16)
            at org.bukkit.craftbukkit.v1_14_R1.CraftWorld.getChunkAt(CraftWorld.java:381)
            at org.bukkit.craftbukkit.v1_14_R1.CraftWorld.getChunkAt(CraftWorld.java:948)
            at org.bukkit.Location.getChunk(Location.java:123)
            at de.myftb.worldmanagement.WorldManagement.loadWorld(WorldManagement.java:360)
            at de.myftb.worldmanagement.WorldManagement.loadWorld(WorldManagement.java:290)
            at de.myftb.worldmanagement.WorldManagement.loadWorlds(WorldManagement.java:397)
            at de.myftb.worldmanagement.WorldManagement.onEnable(WorldManagement.java:493)
            at org.bukkit.plugin.java.JavaPlugin.setEnabled(JavaPlugin.java:263)
            at org.bukkit.plugin.java.JavaPluginLoader.enablePlugin(JavaPluginLoader.java:338)
            at org.bukkit.plugin.SimplePluginManager.enablePlugin(SimplePluginManager.java:420)
            - locked <0x00000002cad32e60> (a org.bukkit.plugin.SimplePluginManager)
            at org.bukkit.craftbukkit.v1_14_R1.CraftServer.enablePlugin(CraftServer.java:467)
            at org.bukkit.craftbukkit.v1_14_R1.CraftServer.enablePlugins(CraftServer.java:381)
            at net.minecraft.server.v1_14_R1.MinecraftServer.a(MinecraftServer.java:477)
            at net.minecraft.server.v1_14_R1.DedicatedServer.init(DedicatedServer.java:290)
            at net.minecraft.server.v1_14_R1.MinecraftServer.run(MinecraftServer.java:873)
            at java.lang.Thread.run(Thread.java:748)
</details>

### Plugin list:
`AnvilColors, ArmorStandEditor, CoreProtect, DragonEggDistribution, Essentials, Hypnos, IpNotifier, IronElevators, KeepItems, Kits, MapLock, mcMMO, Ontime, OpenInv, PermissionsEx, PersonalFlight, PlayerReports, QuickShop, RandomItems, RestartOnCrash, RFChairs, ServerSigns, Shopkeepers, SimpleAPI, SimplePets, TeamspeakBridge, TPSCollector, ValidationSystem, VanishNoPacket, Vault, VoidWorld, WorldEdit, WorldGuard, WorldManagement`

### Actions to reproduce (if known):
Unfortunately I'm currently unable to give any actions to reproduce as the freezes occur quite randomly and I wasn't able to track down a specific cause to this issue yet.

### Paper build number:
```
This server is running Paper version git-Paper-aa022fda (MC: 1.14.4) (Implementing API version 1.14.4-R0.1-SNAPSHOT)
Checking version, please wait...
Previous version: git-Paper-"9fe63a16" (MC: 1.14.4)
You are 5 version(s) behind
```

**Comment by Spottedleaf** [comment]
please provide a watchdog report (it will contain a thread dump itself and incremental thread dumps of the main thread)

~~also 9fe63a16 is the first build of 1.14.4, please update~~ oops read the version incorrectly

**Comment by dogboy21** [comment]
Unfortunately I didn't get an output from the watchdog on todays freeze.
I'll post a report as soon as I can get my hands on one.

**Comment by Spottedleaf** [comment]
check the watchdog config (i.e not disabled, check startup params since paper let's you disable it)
if the watchdog is not running then there's a much bigger issue at play here

**Comment by dogboy21** [comment]
A few minutes ago, the server froze again. This time I got a watchdog report: https://gist.github.com/dogboy21/c968d7e264b54fea363ac1d554785786

**Comment by Spottedleaf** [comment]
It looks like players moving may be causing chunk loads, can you try to configure `prevent-moving-into-unloaded-chunks` to `true` in paper.yml?

**Comment by dogboy21** [comment]
It seems like this was the cause for the last watchdog report, although we already had many different causes to this issue. But I'll still try the mentioned config.

In the meantime I've tested one of the latest changes to your async chunk IO pr (https://github.com/Spottedleaf/Concrete-PR/commit/73172ed88797d96a9486f167730ba677b74e46bb) and it seems like the problem is gone. We might just have been lucky, but we didn't have a single freeze today.

**Comment by dogboy21** [comment]
The server froze again, even with `prevent-moving-into-unloaded-chunks` set to `true`.
A second server with the async chunk pr still hasn't crashed yet.

**Comment by Phineas** [comment]
Had the same issue as you, with TPS fluctuations as well. I just built the chunk pr version and I still get weird TPS fluctuations. Seemed fine when I was testing with 5 players, but now it's on the main server there seems to be a big difference.

https://timings.aikar.co/?id=63dcf47c4f324eb6a5ef11ba518a3425

**Comment by Spottedleaf** [comment]
Luckily for you async was just merged into the main branch

Please update, if this occurs again the watchdog logs will give a little more detail in this situation

**Comment by Phineas** [comment]
I’ve updated, still having the major TPS issues - sometimes goes down to 2-4 TPS when there’s only 20-30 online. There’s 230GB if memory dedicated to the server, the CPU is an E5-Xeon

**Comment by LoneDev6** [comment]
how many cores your CPU has reserved to your server?

**Comment by Phineas** [comment]
@LoneDev6 There’s 32 cores in the entire dedi

**Comment by danny6167** [comment]
Have you tried not allocating a ludicrous amount of memory?
230G is significantly higher than the max of 10GB suggested by aikar.
https://aikar.co/2018/07/02/tuning-the-jvm-g1gc-garbage-collector-flags-for-minecraft/

**Comment by dogboy21** [comment]
It seems like the deadlock in ChunkProviderServer is mostly gone.
Currently we're experiencing frequent freezes due to another cause.
Watchdog Report: https://gist.github.com/dogboy21/148aa576b808b94e9205481160f2788a

**Comment by Spottedleaf** [comment]
we have changed a little bit of the loader system since your build, can you try latest?

**Comment by dogboy21** [comment]
I've kept an eye on the server the last weeks.
The problem still exists, although it's happening less frequently now.

**Comment by Spottedleaf** [comment]
Any logs?

**Comment by dogboy21** [comment]
https://gist.github.com/dogboy21/77d2c1abc31e6b85fc69902b633b5f05
https://gist.github.com/dogboy21/4f700fd52b599373a1c1b3b1a8c9d0db

**Comment by Spottedleaf** [comment]
1st log:
the logic for loading spawn is a bit weird, I'm going to change it to make it safer since it can call load events, and plugins can change how many chunks get loaded...

2nd log:
It's possible that generation took a long time but I'll add in more debug info in a crash to indicate whether that's the case

**Comment by dogboy21** [comment]
Watchdog Report of 2 freezes with #2595:

https://gist.github.com/dogboy21/d574330efb147f67dad6f72366656f52
https://gist.github.com/dogboy21/060389850a20ae096ac0eb8acff76046

**Comment by Chew** [comment]
As this is an old issue, and 1.14 being 1.14, I'm closing this issue. If you're still experiencing this after 2 major updates and a ton of patches, feel free to open a new issue. Thanks for the report!
